src/owr/OptimisticWithdrawalRecipientPectra.sol

```diff
@@ -0,0 +1,411 @@
+    unchecked {
+      // shouldn't overflow
+      return _getArgUint256(TRANCHES_OFFSET + i * ONE_WORD);
+    }
+  }
+
+  /// -----------------------------------------------------------------------
+  /// storage - mutables
+  /// -----------------------------------------------------------------------
+  /// @dev set to `true` after owner is initialized
+  bool public initialized;
+
+  /// Amount of active balance set aside for pulls
+  /// @dev ERC20s with very large decimals may overflow & cause issues
+  uint128 public fundsPendingWithdrawal;
+
+  /// Amount of distributed OWRecipient token for principal
+  /// @dev Would be less than or equal to amount of stake
+  /// @dev ERC20s with very large decimals may overflow & cause issues
+  uint128 public claimedPrincipalFunds;
+
+  /// Mapping to account balances for pulling
+  mapping(address => uint256) internal pullBalances;
+
+  /// -----------------------------------------------------------------------
+  /// constructor
+  /// -----------------------------------------------------------------------
+
+  // solhint-disable-next-line no-empty-blocks
+  /// clone implementation doesn't use constructor
+  constructor(address _pectraWithdrawalAddress, address _pectraConsolidationAddress) {
+    pectraConsolidationAddress = _pectraConsolidationAddress;
+    pectraWithdrawalAddress = _pectraWithdrawalAddress;
+  }
+
+  /// -----------------------------------------------------------------------
+  /// functions
+  /// -----------------------------------------------------------------------
+
+  /// -----------------------------------------------------------------------
+  /// functions - public & external
+  /// -----------------------------------------------------------------------
+
+  /// @dev initializes the owner
+  /// @param _owner the owner address
+  function initialize(address _owner) public {
+    if (initialized) revert InvalidRequest_Initialize();
+    _initializeOwner(_owner);
+    initialized = true;
+  }
+
+  /// emit event when receiving ETH
+  /// @dev implemented w/i clone bytecode
+  /* receive() external payable { */
+  /*     emit ReceiveETH(msg.value); */
+  /* } */
+
+  function requestConsolidation(bytes memory source, bytes memory target) external payable onlyOwnerOrRoles(CONSOLIDATION_WITHDRAWAL_ROLE) {
+    if (source.length != 48 || target.length != 48) revert InvalidConsolidation_Failed();
+
+    // ;; Input data has the following layout:
+    // ;;
+    // ;;  +--------+--------+
+    // ;;  | source | target |
+    // ;;  +--------+--------+
+    // ;;      48       48
+
+    (bool ret,) = pectraConsolidationAddress.call{value: msg.value}(abi.encodePacked(source, target));
+    if (!ret) revert InvalidConsolidation_Failed();
+    emit ConsolidationRequested(msg.sender);
+  }
+
+  /// Requests principal withdrawal
+  function requestPrincipalWithdrawal(bytes memory pubkey, uint8 amount) external payable onlyOwnerOrRoles(PRINCIPAL_WITHDRAWAL_ROLE) {
```

Comment:
should this be an array given that multiple validators will use a OWR?

Reply:
I can add a batchRequest method as well. For both, principal and rewards